Overhaul the Gradient Editor

MewPurPur opened the pull request:
Closes https://github.com/godotengine/godot-proposals/issues/6555

(I murdered this PR twice with bad rebasing, sorry about that :/   replaces #71746)

I *just* wanted to implement snapping in the Gradient Editor. But when I pulled on that thread, the whole ball of yarn unraveled.

### Refactoring

Previously, we had two sets of files (`gradient_editor` and `gradient_editor_plugin`).

`gradient_editor` had the GradientEditor class. This was for the colorful rectangle that lets you edit a gradient, having all the logic around GUI input on it and also the color selector.

`gradient_editor_plugin` had the EditorInspectorPluginGradient class, which included the GradientEditor and a small button for reversing the gradient. It also had the GradientReverseButton class, because... yeah, needed a hack just to add that button. This wasn't scalable.

I renamed GradientEditor to GradientEdit, whereas the whole thing with the buttons is now a GradientEditor. These were all moved to one file like other editor plugins.

| Before | After |
| ------------- | ------------- |
| ![image](https://user-images.githubusercontent.com/85438892/214057023-998248b9-edc2-4d74-9638-930fe120ea75.png) | ![image](https://user-images.githubusercontent.com/85438892/213755937-c378c101-97ed-44f9-ad43-a845e14669ac.png) |

### Snap Button

Looks like this:

![image](https://user-images.githubusercontent.com/85438892/213762435-30277c4c-1bf4-445e-9a09-72748e359bda.png)

The slider and the vertical lines are only shown when snap is toggled.

I removed `Ctrl+Shift+Drag` (which snaps with 40 partitions) as the snap button makes this quite obsolete. `Ctrl+Drag` still works though, but it applies the snap count of the gradient.

`Shift+Drag` still works and is quite useful to get two points near each other when snapping. I gave it a smaller threshold by default. This can also be combined with `Ctrl`.

### Improved Undo/Redo system

The undo/redo system wasn't functional. The single manager in `_ramp_changed()` left thousands of vague "Gradient Changed" messages in the output and would usually only lead you one step back before breaking, and that step back would usually not be where you wanted to get.

I didn't want to leave this system even more broken than before, so I modernized it. Now it has more descriptive messages and is more predictable. Code was simplified a lot too, removing some unused or unnecessary things.

**Testing/Review on this would be appreciated because it might corrupt gradients if not done right!**

### Misc enhancements

* Improved the condition for when the handle is drawn in black or white, to maximize its contrast.

| Before | After |
| ------------- | ------------- |
| ![image](https://user-images.githubusercontent.com/85438892/214057514-b377373d-3a94-49c4-a749-b0a35ccac814.png) | ![image](https://user-images.githubusercontent.com/85438892/213759303-d2b3fb6b-542c-4110-ab94-5a1e99255530.png) |

* Added an indicator for overbright colors. (as shown in the previous section)
* Improved the condition for whether the color picker is shown above or below. If the color picker can fit neither below nor above, then it picks the direction with more space.
* Added hover indicators:
![image](https://user-images.githubusercontent.com/85438892/213757502-d371ffbd-9be9-4a2b-ae71-edca487abad6.png)
* Added the ability to deselect all handles by right-clicking anywhere without handles.
* Fixed drawing of translucent colors by using a separate checkerboard pattern for each handle.
* Added solid color indicators to translucent color handles:
![image](https://github.com/godotengine/godot/assets/85438892/06e4e364-6f1c-4f79-a1c2-68366a64ac9f)
* Improved the logic for keeping handles selected during operations (i.e. flipping when the gradient is reversed).
* Visually glitched handles with offsets outside of the [0, 1] range are no longer shown.
* Fixed typo in `gradient.h` (ponit -> point :carousel_horse:)

MewPurPur committed:
Overhaul the Gradient Editor

KoBeWi commented:
Tested the editor and functionally it's good, although it would be nice if snapping was remembered. Right now it resets even if you just fold the gradient. You could use project metadata for that, or a static member.

KoBeWi commented on editor/plugins/gradient_editor_plugin.h:
```diff
@@ -33,25 +33,99 @@
 #include "editor/editor_inspector.h"
 #include "editor/editor_plugin.h"
-#include "gradient_editor.h"
+#include "editor/editor_spin_slider.h"
```
Do not add new includes in the header if they are not necessary. Use forward declarations instead.

MewPurPur replied:
Alrighty. BTW did I get the whole "public / protected / private" thing right in the header?

MewPurPur commented:
Not sure right now how I could save snapping, especially since it seems mostly useful on a per-gradient basis... GradientTexture2D's snapping could also use the solution to this.

KoBeWi commented on editor/plugins/gradient_editor_plugin.h:
```diff
@@ -33,25 +33,100 @@
 #include "editor/editor_inspector.h"
 #include "editor/editor_plugin.h"
-#include "gradient_editor.h"
+#include "scene/gui/color_picker.h"
+#include "scene/gui/popup.h"
```
The other includes can be moved too 🙃
Includes in header are most often needed only if you inherit some class. Here only `editor_inspector.h` and `editor_plugin.h` are needed (they indirectly include other classes).

KoBeWi commented on editor/plugins/gradient_editor_plugin.cpp:
```diff
@@ -31,27 +31,583 @@
+int GradientEditorRect::get_point_at(int x) {
+	int result = -1;
+	int total_w = get_gradient_rect_width();
+	float min_distance = 1e20;
```
```suggestion
	float min_distance = handle_width * 0.8;
```
Allows `min` to be removed further down.

KoBeWi commented on editor/plugins/gradient_editor_plugin.cpp:
```diff
@@ -31,27 +31,583 @@
+
+	// Determine in which direction to show the popup. By default popup below.
+	// But if the popup doesn't fit below and the Gradient Editor is in the bottom half of the viewport, show above.
+	bool show_above = false;
+	if (get_global_position().y + get_size().y + minsize.y > viewport_height && get_global_position().y * 2 + get_size().y > viewport_height) {
+		show_above = true;
+	}
```
Can be simplified to
```suggestion
	bool show_above = (get_global_position().y + get_size().y + minsize.y > viewport_height && get_global_position().y * 2 + get_size().y > viewport_height);
```

KoBeWi commented on editor/plugins/gradient_editor_plugin.cpp:
```diff
@@ -31,27 +31,583 @@
+
+void GradientEditor::_reverse_button_pressed() {
+	EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
+	undo_redo->create_action(TTR("Reverse Gradient"));
```
```suggestion
	undo_redo->create_action(TTR("Reverse Gradient"), UndoRedo::MERGE_DISABLE, gradient.ptr());
```
Required to properly use scene local history.

MewPurPur replied:
Hmm, something is telling me this might result in problems, but let me try.

MewPurPur replied:
Well, GradientEditor is not the one who holds the gradient. So I put this in another `_reverse_gradient` func in GradientEditorRect, please take a look to make sure it's right as tbh I don't understand the problem

KoBeWi replied:
Looks good now. Previously if you had built-in Gradient on a scene, reverse would register as global history action instead of scene.

KoBeWi commented on editor/plugins/gradient_editor_plugin.cpp:
```diff
@@ -31,27 +31,583 @@
+	add_child(toolbar);
+
+	reverse_button = memnew(Button);
+	reverse_button->set_tooltip_text(TTR("Reverse/mirror gradient"));
```
```suggestion
	reverse_button->set_tooltip_text(TTR("Reverse/Mirror Gradient"));
```

KoBeWi commented on editor/plugins/gradient_editor_plugin.cpp:
```diff
@@ -31,27 +31,585 @@
+///////////////////////
+
+void GradientEditor::_reverse_button_pressed() {
+	gradient_editor_rect->_reverse_gradient();
```
You could connect the signal to this method directly. `_reverse_button_pressed()` is no longer needed.

MewPurPur commented:
Not safe to merge at the moment! I'll delete this when it's safe.

MewPurPur commented:
Reverted the last change KoBeWi suggested, it makes gradient reversing often crash for some reason.

YuriSizov commented:
Just so we're clear, there is no proposal for the additions that this PR makes? Refactoring is good, and is probably welcome on its own, but what's the justification for the snap feature, for instance?

MewPurPur commented:
It's meant as an overhaul to make the class look better, and the PR is 90% just that. Before I started working on the gradient preview, it looked like this, is a proposal really needed to say that it looks underbaked?

![image](https://user-images.githubusercontent.com/85438892/226437941-929a6c0b-a3d7-4615-80c8-af92e43adac8.png)

We already have snapping, but it's a hidden Ctrl+Drag trick and it's not configurable, so this PR simply adds a way to adjust that. I've seen a few people say they miss the feature in various places, but just no one has come forward to actually propose it. I'll do so if needed.

YuriSizov commented:
> is a proposal really needed to say that it looks underbaked?

Well, I asked for a proposal regarding the functional changes, not the refactoring. You don't have to split the PR, but a formal proposal for the snapping feature would be good to have. Problem with people asking for it all over the place is that at the moment that request is invisible, and thus we don't even know if this solution improves the flow for them.

MewPurPur commented:
https://github.com/godotengine/godot-proposals/issues/6555

MewPurPur commented:
> Right now it resets even if you just fold the gradient. You could use project metadata for that, or a static member.

@KoBeWi I made each gradient remember its own snapping settings by putting metadata in it. Sound good?

KoBeWi commented:
Isn't metadata saved with the gradient? 🤔

MewPurPur commented:
Yeah, I think it's the most useful when remembered on a per-gradient basis.

KoBeWi commented:
> Hmm, set_value() doesn't seem to trigger the value_changed signal for some reason actually...

It emits the signal only when the value changes. Also make sure it's connected to the correct object.

MewPurPur commented:
Ah, looks like it doesn't happen until, hmm, _ready()? Might be missing something here but I'm pretty surprised.

MewPurPur commented:
> Reverted the last change KoBeWi suggested, it made gradient reversing often crash on my machine for some reason. Something about this operation is hanging on loose strings, but I just can't figure it out.

#75234

So the operation was broken in the first place, not my fault. I'll need to rebase when this is fixed to remove the weird workarounds to this I made.

MewPurPur commented:
Fixed :)  Reversing should be good now.

MewPurPur commented on editor/plugins/gradient_editor_plugin.cpp:
```diff
@@ -31,27 +31,590 @@
+			reverse_button->set_icon(get_theme_icon(SNAME("ReverseGradient"), SNAME("EditorIcons")));
+			snap_button->set_icon(get_theme_icon(SNAME("SnapGrid"), SNAME("EditorIcons")));
+		} break;
+		case NOTIFICATION_READY: {
+			Ref<Gradient> gradient = gradient_editor_rect->get_gradient();
+			// Set snapping settings based on the gradient.
+			if ((bool)gradient->get_meta(SNAME("_snap_enabled"), false) == true) {
+				snap_button->set_pressed(true);
+			}
+			if (gradient->has_meta(SNAME("_snap_count"))) {
+				snap_count_edit->set_value(gradient->get_meta(SNAME("_snap_count")));
+			}
+		} break;
```
Adjusts the toggle button and the spinbox's value based on the gradient metadata.

I has to do this in READY, because otherwise the the `value_changed` signal on the spinbox wouldn't emit.

MewPurPur commented on editor/plugins/gradient_editor_plugin.cpp:
```diff
@@ -31,27 +31,590 @@
+	gradient->connect("changed", callable_mp((CanvasItem *)this, &CanvasItem::queue_redraw));
+}
+
+Ref<Gradient> GradientEditorRect::get_gradient() {
+	return gradient;
+}
```
The GradientEditor class needed the gradient's metadata to update its buttons, but `set_gradient()` was too early. So I'm calling this at NOTIFICATION_READY.

Only thought of it now, but it might be better to instead store this metadata in two global variables?

KoBeWi replied:
What do you mean by global variables?

MewPurPur replied:
I thought of having a `snap_count` variable in GradientEditor too, which gets set in `set_gradient()` and then reused in `NOTIFICATION_READY`. On a second though, a variable that only gets used the before the control is ready doesn't sound good.

MewPurPur reviewed (commented):
Some comments to help isolate the metadata implementation for easier reviewing.

KoBeWi commented:
I'm not really fan of storing editor-specific metadata, but we don't really have any better way to handle this.
Although I think you could only store the metadata if it's non-default value.
Also I wonder if saving disabled snap is even necessary. You could e.g. only store snap count and if the snap is disabled, don't store it. So the editor would enable snap if the `snap_count` value exists.

KoBeWi commented on editor/plugins/gradient_editor_plugin.cpp:
```diff
@@ -31,27 +31,590 @@
+		case NOTIFICATION_READY: {
+			Ref<Gradient> gradient = gradient_editor_rect->get_gradient();
+			// Set snapping settings based on the gradient.
+			if ((bool)gradient->get_meta(SNAME("_snap_enabled"), false) == true) {
```
```suggestion
			if ((bool)gradient->get_meta(SNAME("_snap_enabled"), false)) {
```
😬

MewPurPur commented:
> btw is the meta the only thing that changed since my last review?

@KoBeWi I also redid gradient reversing since `Gradienrt.reverse()` got fixed and I no longer needed a hack to make it work.

MewPurPur commented:
Sure thing removing the meta when it's default data. The latter I don't think is a good idea as the number of snap divisions would get lost if you close the snap button and fold the editor.

KoBeWi commented on editor/plugins/gradient_editor_plugin.cpp:
```diff
@@ -31,27 +31,594 @@
+void GradientEditorRect::set_snap_enabled(bool p_enabled) {
+	snap_enabled = p_enabled;
+	queue_redraw();
+	if (!gradient.is_null()) {
```
```suggestion
	if (gradient.is_valid()) {
```
Looks better.

KoBeWi commented:
> Removing it I don't think is a good idea because the number of snap divisions would get lost if you untoggle the snap button and fold the editor.

But is it necessary if it has default value?

KoBeWi commented:
Yes.

Geometror reviewed (changes requested):
Great work!
Tested it a bit and it feels very nice to work with, the code looks fine too.
However, there are still some things to consider:

- > I removed Ctrl+Drag (which snaps with 10 partitions) and Ctrl+Shift+Drag (which snaps with 40 partitions) as the snap button makes these tricks quite obsolete.

	As with the curve editor overhaul, I am not a fan of this, because it's a really useful (and fast) way to edit your gradient. I don't think the snap button makes this function obsolete since you have to click on it each time to toggle snapping - moving away from the gradient, which slows you down. I'd suggest adding this function back in a way that it respects the snap division count (e.g. Ctrl + Drag snap_count steps, Ctrl + Shift + Drag snap_count*4 steps).

- I know, it's quite obvious what it does, but maybe add a Label (in front of) or Tooltip to the snap count spin slider for better accessibility and comprehensibility

MewPurPur commented:
>  I'd suggest adding this function back in a way that it respects the snap division count

Doned. It respects the snap division count and also draws the grid temporarily. I didn't re-implement Ctrl+Shift though.

I initially didn't want to remove it, but I thought the best way to implement it would be to temporarily toggle the button so that the division count becomes visible. I couldn't do that so decided to gut it.

> I know, it's quite obvious what it does, but maybe add a Label (in front of) or Tooltip to the snap count spin slider for better accessibility and comprehensibility

Maybe in a follow-up? I am not sure what you mean here, and the editor's spinboxes I believe are hard-coded to just show their value, and what the key modifiers do for they are sliders.

MewPurPur commented:
- Fixed errors that popped up if you clicked on the color "button" without a selected point.
- Improved the "draw the handle in black or white" condition further by accounting for transparent colors.
- Added a functionality for right-clicking while moving a handle to bring it back to its old position (similar to moving objects in the canvas editor)

These are all pretty isolated, there's no way they broke anything new.

KoBeWi commented:
I tested and glanced at the code after recent changes and the PR still looks ok.

> Maybe in a follow-up? I am not sure what you mean here, and the editor spinboxes' tooltips I believe are hard-coded to just show their value, and what the key modifiers do for sliders.

I think he meant something like this
![image](https://github.com/godotengine/godot/assets/2223172/b9c34732-b7db-4a88-9a3c-12cb820bde1f)
IMO it's not really necessary, and wouldn't be consistent. The newly added snapping in TileSet doesn't have it.

> Ctrl + Shift + Drag snap_count*4 steps

This seems unaddressed. idk if it's that important.

MewPurPur commented:
> This seems unaddressed. idk if it's that important.

I think there's a misunderstanding, I didn't implement it because the two buttons have separate functionalities. SHIFT allows you to snap a handle right next to another, which is useful for sharply changing the color. I think this is especially useful with snapping, so I didn't want to make CTRL+SHIFT do something different.

Calinou reviewed (approved):
Tested locally, it works. Great job :slightly_smiling_face: 

Code looks good to me at a glance.

MewPurPur commented:
Fixed again, yee.

Made points able to be deselected if you RMB on empty space as the selection can be annoying and obstruct the color inside.

MewPurPur commented:
![image](https://github.com/godotengine/godot/assets/85438892/06e4e364-6f1c-4f79-a1c2-68366a64ac9f)

There was something of an agreement on Rocketchat to go with also adding these little indicators to make the tint of transparent colors clearer.

AThousandShips commented on editor/plugins/gradient_editor_plugin.h:
```diff
@@ -33,25 +33,106 @@
+	void _color_changed(const Color &p_color);
+	void _redraw();
+
+	int get_point_at(int x_pos);
```
```suggestion
	int get_point_at(int p_x_pos);
```

AThousandShips commented on editor/plugins/gradient_editor_plugin.cpp:
```diff
@@ -31,27 +31,625 @@
+#include "scene/gui/separator.h"
 
-void GradientReverseButton::_notification(int p_what) {
+int GradientEdit::get_point_at(int x) {
```
```suggestion
int GradientEdit::get_point_at(int p_x_pos) {
```

AThousandShips commented on editor/plugins/gradient_editor_plugin.cpp:
```diff
@@ -31,27 +31,625 @@
+			break;
+		}
+		// Check if we clicked at point.
+		float distance = ABS(x - gradient->get_offset(i) * total_w);
```
```suggestion
		float distance = ABS(p_x_pos - gradient->get_offset(i) * total_w);
```

MewPurPur commented:
Sorry about these bunch of pushes, I pushed a small fix from a wrong account

MewPurPur commented:
Will this get into 4.2? I think it's super polished now.

YuriSizov commented:
I'd like @Geometror to give it another look since he previously requested changes.

Geometror commented on editor/plugins/gradient_editor_plugin.h:
```diff
@@ -33,25 +33,106 @@
+	ColorPicker *picker = nullptr;
+
+	bool snap_enabled = false;
+	float snap_count = 10;
```
Should be an integer type (unless there is a reason for it which I overlooked)

MewPurPur replied:
Thanks, I also noticed for some reason I made the hovered point's index a float. So fixed that too.

Geometror reviewed (commented):
> Maybe in a follow-up? I am not sure what you mean here, and the editor spinboxes' tooltips I believe are hard-coded to just show their value, and what the key modifiers do for sliders.

What I meant was something like this:

![grafik](https://github.com/godotengine/godot/assets/50084500/33bbb886-07dd-43b4-8825-1811be829c25)

I know it's quite obvious what it does, but I think adding such a label makes its purpose clearer IMO and improves accessibility (e.g. for people dependent on a screen reader)

MewPurPur commented:
![image](https://github.com/godotengine/godot/assets/85438892/4c60783e-3978-4cad-a4fb-ff3677ac72b9)

^ Still didn't implement a label, I'm mostly just not sure how to handle its size when the inspector is very squished, like mine is usually.

MewPurPur commented:
This PR is fucking cursed

AThousandShips commented on editor/plugins/gradient_editor_plugin.cpp:
```diff
@@ -31,27 +31,626 @@
+	int half_handle_width = handle_width * 0.5;
+
+	// Draw gradient.
+	draw_texture_rect(get_theme_icon(SNAME("GuiMiniCheckerboard"), SNAME("EditorIcons")), Rect2(half_handle_width, 0, total_w, h), true);
```
```suggestion
	draw_texture_rect(get_editor_theme_icon(SNAME("GuiMiniCheckerboard")), Rect2(half_handle_width, 0, total_w, h), true);
```
Dedicated system

AThousandShips commented on editor/plugins/gradient_editor_plugin.cpp:
```diff
@@ -31,27 +31,626 @@
+
+		if (inside_col.a < 1) {
+			// If the color is translucent, draw a little opaque rectangle at the bottom to more easily see it.
+			draw_texture_rect(get_theme_icon(SNAME("GuiMiniCheckerboard"), SNAME("EditorIcons")), rect, true);
```
```suggestion
			draw_texture_rect(get_editor_theme_icon(SNAME("GuiMiniCheckerboard")), rect, true);
```

AThousandShips commented on editor/plugins/gradient_editor_plugin.cpp:
```diff
@@ -31,27 +31,626 @@
+	if (selected_index != -1) {
+		Color grabbed_col = gradient->get_color(selected_index);
+		if (grabbed_col.a < 1) {
+			draw_texture_rect(get_theme_icon(SNAME("GuiMiniCheckerboard"), SNAME("EditorIcons")), Rect2(button_offset, 0, h, h), true);
```
```suggestion
			draw_texture_rect(get_editor_theme_icon(SNAME("GuiMiniCheckerboard")), Rect2(button_offset, 0, h, h), true);
```

AThousandShips commented on editor/plugins/gradient_editor_plugin.cpp:
```diff
@@ -31,27 +31,626 @@
+				draw_line(Vector2(handle_start_x + handle_thickness / 2.0, h * 0.9 - handle_thickness), Vector2(handle_start_x + handle_width - handle_thickness / 2.0, h * 0.9 - handle_thickness), border_col, handle_thickness);
+			}
+			rect = rect.grow(-handle_thickness);
+			const Color focus_col = get_theme_color(SNAME("accent_color"), SNAME("Editor"));
```
```suggestion
			const Color focus_col = get_theme_color(SNAME("accent_color"), EditorStringName(Editor));
```

AThousandShips commented on editor/plugins/gradient_editor_plugin.cpp:
```diff
@@ -31,27 +31,626 @@
+void GradientEditor::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_THEME_CHANGED: {
+			reverse_button->set_icon(get_theme_icon(SNAME("ReverseGradient"), SNAME("EditorIcons")));
```
```suggestion
			reverse_button->set_icon(get_editor_theme_icon(SNAME("ReverseGradient")));
```

AThousandShips commented on editor/plugins/gradient_editor_plugin.cpp:
```diff
@@ -31,27 +31,626 @@
+	switch (p_what) {
+		case NOTIFICATION_THEME_CHANGED: {
+			reverse_button->set_icon(get_theme_icon(SNAME("ReverseGradient"), SNAME("EditorIcons")));
+			snap_button->set_icon(get_theme_icon(SNAME("SnapGrid"), SNAME("EditorIcons")));
```
```suggestion
			snap_button->set_icon(get_editor_theme_icon(SNAME("SnapGrid")));
```

AThousandShips commented on editor/plugins/gradient_editor_plugin.cpp:
```diff
@@ -31,27 +31,626 @@
 #include "editor/editor_node.h"
 #include "editor/editor_scale.h"
 #include "editor/editor_settings.h"
 #include "editor/editor_undo_redo_manager.h"
```
```suggestion
#include "editor/editor_string_names.h"
#include "editor/editor_undo_redo_manager.h"
```

Geometror reviewed (approved):
Ok, I guess it's fine for now without such a Label. Unfortunately, Godot's layout system isn't currently capable of this: 
There's no `preferred_size`/`maximum_size` property, so when you set the text overrun mode to `TextServer::OVERRUN_TRIM_ELLIPSIS` the label has a minimum size of 0. The only option that is currently possible would be to set the h size flags to `SIZE_EXPAND_FILL`,  but then the spin slider gets pushed to the right.

YuriSizov commented on editor/plugins/gradient_editor_plugin.h:
```diff
@@ -33,25 +33,106 @@
 #include "editor/editor_inspector.h"
 #include "editor/editor_plugin.h"
-#include "gradient_editor.h"
+#include "scene/resources/gradient_texture.h"
```
The include is redundant here since you forward declare the class. Has to be moved to the cpp file.

YuriSizov commented on editor/plugins/gradient_editor_plugin.h:
```diff
@@ -33,25 +33,106 @@
+	};
+
+	GrabMode grabbing = GRAB_NONE;
+	float pre_grab_offset;
+	int pre_grab_index;
```
Need to be initialized.

YuriSizov commented on editor/plugins/gradient_editor_plugin.h:
```diff
@@ -33,25 +33,106 @@
+	void remove_point(int p_index);
+	void set_offset(int p_index, float p_offset);
+	void set_color(int p_index, const Color &p_color);
+	void reverse_gradient();
+
+	void _reverse_gradient();
```
Doesn't look like the public method is implemented, and the underscored method doesn't need to be public.

MewPurPur replied:
Yeah, it's an artifact of the reverse() method straight up not working because it forgot to sort. Now it can be called outright without any extra steps.

YuriSizov commented on editor/plugins/gradient_editor_plugin.h:
```diff
@@ -33,25 +33,106 @@
+
+public:
+	void set_gradient(const Ref<Gradient> &p_gradient);
+	Ref<Gradient> get_gradient();
+
+	ColorPicker *get_picker();
+	PopupPanel *get_popup();
```
Should be const.

YuriSizov commented on editor/plugins/gradient_editor_plugin.h:
```diff
@@ -33,25 +33,106 @@
+	int draw_spacing = BASE_SPACING;
+	int handle_width = BASE_HANDLE_WIDTH;
+
+	int get_gradient_rect_width();
+
+	void _color_changed(const Color &p_color);
+	void _redraw();
+
+	int get_point_at(int p_xpos);
+	int predict_insertion_index(float p_offset);
+	void show_color_picker();
```
There are a few private methods without underscores. Also getters should be const.

YuriSizov commented on editor/plugins/gradient_editor_plugin.h:
```diff
@@ -33,25 +33,106 @@
+	int hovered_index = -1;
+
+	// Make sure to use the scaled values below.
+	const int BASE_SPACING = 4;
+	const int BASE_HANDLE_WIDTH = 8;
```
Should these be static?

MewPurPur replied:
Not sure, I followed the old conventions in gradient_editor.cpp

YuriSizov replied:
I think you can make them static, since there is another constant in the next class that is. I don't see why we'd want to have a non-static const here.

YuriSizov commented on editor/plugins/gradient_editor_plugin.cpp:
```diff
@@ -31,27 +31,627 @@
+
+	preview_texture.instantiate();
+	preview_texture->set_width(1024);
+	add_child(popup, false, INTERNAL_MODE_FRONT);
```
Why is it here and not next to the node allocating code?

MewPurPur replied:
I think I didn't change this code, I assume it was just like "Setting up everything before adding it" sorta thing?

YuriSizov replied:
For style consistency it's better to move it to where we memnew the node, I think.

YuriSizov commented on editor/plugins/gradient_editor_plugin.cpp:
```diff
@@ -31,27 +31,627 @@
+void GradientEdit::_notification(int p_what) {
 	switch (p_what) {
+		case NOTIFICATION_ENTER_TREE: {
+			picker->connect("color_changed", callable_mp(this, &GradientEdit::_color_changed));
```
This should be in the constructor, or is there a reason for this approach?

MewPurPur replied:
Not sure, also copied code from the old class. Maybe signals connected in a constructor are kinda inconsistent? I remember something being more inconsistent in constructors than enter_tree, but probably not this, so I'll move this.

YuriSizov replied:
You can't use themes in the constructor, and you also may have issues interacting with the node being constructed itself. But connecting to signals of children should be completely fine.

YuriSizov commented on editor/plugins/gradient_editor_plugin.cpp:
```diff
@@ -31,27 +31,627 @@
+}
+
+Size2 GradientEdit::get_minimum_size() const {
+	return Size2(0, 60) * EDSCALE;
```
Is there a reason to override the method instead of setting custom minimum size in the constructor?

MewPurPur replied:
Not sure, actually I didn't write this. A lot of it is just from merging gradient_editor.cpp and gradient_editor_plugin.cpp. I only tweaked the height a bit to accommodate the better-looking buttons. I mostly avoided touching parts of the code I don't entirely understand.

MewPurPur replied:
If you suggest that overriding the method is inferior, maybe I could clean up declarations like these? I think I remember this being on the CurveEditor too, and it's probably in a lot more places.

YuriSizov replied:
I don't think anything would break if you used custom minimum size. Can you give it a try?

YuriSizov replied:
> If you suggest that overriding the method is inferior, maybe I could clean up declarations like these in a follow-up PR? I think I remember this being on the CurveEditor too, and it's probably in a lot more places.

I don't remember there being many such places, but if you're too concerned about the implications, this can be set aside for now.

MewPurPur commented:
Reversing broke again if you spam-click the button. Help

YuriSizov commented:
@MewPurPur How does it break? Have you considered adding a guard to the method, so it doesn't run again until it finishes the previous call?

MewPurPur commented:
No, not intentional, sorry, I reset my fix-up thinking that was what caused the glitch. I'll also rebase to be sure.

- Spam-clicking reverse causes a tiny lag spike and points of the gradient occasionally disappear
- Doesn't happen when I just spam reverse() on a gradient in my project
- Undo also causes it (but the Undo of reverse() is also reverse() so that's not surprising)

Idek how to put guards in Undo Redo, but I'll try more things in a little.

YuriSizov commented:
For UndoRedo you can use the merge events feature, then multiple subsequent actions should be treated as one operation. But I meant adding guards to the method that handles the click itself, before the UndoRedo gets called.

MewPurPur commented:
Okay, now all the resolved things are fixed, sorry for the blunder. The reverse bug isn't fixed though, and it's easier to trigger than I thought. You don't have to spam-click, I had it happen after spacing out my clicks for about a second. (though it never happens if you hold off for longer than a second)

YuriSizov commented on editor/plugins/gradient_editor_plugin.h:
```diff
@@ -33,26 +33,102 @@
+
+public:
+	void set_gradient(const Ref<Gradient> &p_gradient);
+	const Ref<Gradient> get_gradient();
+
+	const ColorPicker *get_picker();
+	const PopupPanel *get_popup();
```
`const` goes at the end of the method signature to indicate that it's no mutating.

akien-mga replied:
If both are wanted, this can possibly be `const PopupPanel *get_popup() const;`.

BTW the `get_gradient()` one should likely be `const Ref<Gradient> &`.

MewPurPur replied:
Oh sorry, I honestly wasn't sure what you meant (and what I was doing :sweat:).

MewPurPur replied:
Can I get the exact syntax I need for the Ref<Gradient> line and for the one in the cpp if it needs changing, plz?

YuriSizov replied:
It can be `const Ref<Gradient> &get_gradient() const;`

YuriSizov commented on editor/plugins/gradient_editor_plugin.h:
```diff
@@ -33,26 +33,102 @@
+	int draw_spacing = BASE_SPACING;
+	int handle_width = BASE_HANDLE_WIDTH;
+
+	int _get_gradient_rect_width();
```
Missing `const` here.

MewPurPur replied:
`int _get_gradient_rect_width() const;` ?

YuriSizov replied:
Yes!

YuriSizov commented on editor/plugins/gradient_editor_plugin.h:
```diff
@@ -33,26 +33,102 @@
+	void _color_changed(const Color &p_color);
+	void _redraw();
+
+	int _get_point_at(int p_xpos);
```
And here.

YuriSizov commented on editor/plugins/gradient_editor_plugin.h:
```diff
@@ -33,26 +33,102 @@
+
+	int _get_point_at(int p_xpos);
+	int _predict_insertion_index(float p_offset);
+	void show_color_picker();
```
Still not underscored.

YuriSizov commented:
> The reverse bug isn't fixed though, and it's easier to trigger than I thought. You don't have to spam-click, I had it happen after spacing out my clicks for about a second. (though it never happens if you hold off for longer than a second)

Then you definitely need some guards that would make the `reverse_gradient` call return early if the gradient is currently being updated, and if the editor itself is currently being updated. Basically, at the start of `reverse_gradient` you would set a flag, and only when `GradientEdit` handles the change do you reset it.

MewPurPur commented:
I found more things about the `reverse_gradient()` bug.

- It's actually not caused by spam-clicking at all. Seems to be a random chance it happens, something like 1 in 3 times. But I never had it happen the first time.
- Seems like points are not disappearing, but instead getting caught up in each other and getting the exact same color.

MewPurPur commented:
W. T. F. I was looking more into what the reverse_gradient() bug is exactly, and it just got stuck in a green color (my gradient didn't have a green color) and then crashed Godot. With this trace:

```
[1] /lib/x86_64-linux-gnu/libc.so.6(+0x42520) [0x7f4ace842520] (??:0)
[2] ColorPicker::set_editor_settings(Object*) (/home/volter/Desktop/godot/scene/gui/color_picker.cpp:408)
[3] EditorNode::setup_color_picker(ColorPicker*) (/home/volter/Desktop/godot/editor/editor_node.cpp:3602)
[4] void call_with_variant_args_helper<EditorNode, ColorPicker*, 0ul>(EditorNode*, void (EditorNode::*)(ColorPicker*), Variant const**, Callable::CallError&, IndexSequence<0ul>) (/home/volter/Desktop/godot/./core/variant/binder_common.h:308 (discriminator 4))
[5] void call_with_variant_args<EditorNode, ColorPicker*>(EditorNode*, void (EditorNode::*)(ColorPicker*), Variant const**, int, Callable::CallError&) (/home/volter/Desktop/godot/./core/variant/binder_common.h:418)
[6] CallableCustomMethodPointer<EditorNode, ColorPicker*>::call(Variant const**, int, Variant&, Callable::CallError&) const (/home/volter/Desktop/godot/./core/object/callable_method_pointer.h:105)
[7] Callable::callp(Variant const**, int, Variant&, Callable::CallError&) const (/home/volter/Desktop/godot/core/variant/callable.cpp:50)
[8] CallableCustomBind::call(Variant const**, int, Variant&, Callable::CallError&) const (/home/volter/Desktop/godot/core/variant/callable_bind.cpp:145)
[9] Callable::callp(Variant const**, int, Variant&, Callable::CallError&) const (/home/volter/Desktop/godot/core/variant/callable.cpp:50)
[10] Object::emit_signalp(StringName const&, Variant const**, int) (/home/volter/Desktop/godot/core/object/object.cpp:1114)
[11] Node::emit_signalp(StringName const&, Variant const**, int) (/home/volter/Desktop/godot/scene/main/node.cpp:3587)
[12] Error Object::emit_signal<>(StringName const&) (/home/volter/Desktop/godot/./core/object/object.h:899)
[13] Window::popup(Rect2i const&) (/home/volter/Desktop/godot/scene/main/window.cpp:1696)
[14] GradientEdit::_show_color_picker() (/home/volter/Desktop/godot/editor/plugins/gradient_editor_plugin.cpp:96 (discriminator 4))
[15] GradientEdit::gui_input(Ref<InputEvent> const&) (/home/volter/Desktop/godot/editor/plugins/gradient_editor_plugin.cpp:292)
[16] Control::_call_gui_input(Ref<InputEvent> const&) (/home/volter/Desktop/godot/scene/gui/control.cpp:1809)
[17] Viewport::_gui_call_input(Control*, Ref<InputEvent> const&) (/home/volter/Desktop/godot/scene/main/viewport.cpp:1586)
[18] Viewport::_gui_input_event(Ref<InputEvent>) (/home/volter/Desktop/godot/scene/main/viewport.cpp:1815 (discriminator 3))
[19] Viewport::push_input(Ref<InputEvent> const&, bool) (/home/volter/Desktop/godot/scene/main/viewport.cpp:3179)
[20] Window::_window_input(Ref<InputEvent> const&) (/home/volter/Desktop/godot/scene/main/window.cpp:1547)
[21] void call_with_variant_args_helper<Window, Ref<InputEvent> const&, 0ul>(Window*, void (Window::*)(Ref<InputEvent> const&), Variant const**, Callable::CallError&, IndexSequence<0ul>) (/home/volter/Desktop/godot/./core/variant/binder_common.h:303 (discriminator 4))
[22] void call_with_variant_args<Window, Ref<InputEvent> const&>(Window*, void (Window::*)(Ref<InputEvent> const&), Variant const**, int, Callable::CallError&) (/home/volter/Desktop/godot/./core/variant/binder_common.h:418)
[23] CallableCustomMethodPointer<Window, Ref<InputEvent> const&>::call(Variant const**, int, Variant&, Callable::CallError&) const (/home/volter/Desktop/godot/./core/object/callable_method_pointer.h:105)
[24] Callable::callp(Variant const**, int, Variant&, Callable::CallError&) const (/home/volter/Desktop/godot/core/variant/callable.cpp:50)
[25] DisplayServerX11::_dispatch_input_event(Ref<InputEvent> const&) (/home/volter/Desktop/godot/platform/linuxbsd/x11/display_server_x11.cpp:3779)
[26] DisplayServerX11::_dispatch_input_events(Ref<InputEvent> const&) (/home/volter/Desktop/godot/platform/linuxbsd/x11/display_server_x11.cpp:3749)
[27] Input::_parse_input_event_impl(Ref<InputEvent> const&, bool) (/home/volter/Desktop/godot/core/input/input.cpp:751)
[28] Input::flush_buffered_events() (/home/volter/Desktop/godot/core/input/input.cpp:1013)
[29] DisplayServerX11::process_events() (/home/volter/Desktop/godot/platform/linuxbsd/x11/display_server_x11.cpp:4851)
[30] OS_LinuxBSD::run() (/home/volter/Desktop/godot/platform/linuxbsd/os_linuxbsd.cpp:931)
[31] /home/volter/Desktop/godot/bin/godot.linuxbsd.editor.dev.x86_64(main+0x19f) [0x563ad0a1a268] (/home/volter/Desktop/godot/platform/linuxbsd/godot_linuxbsd.cpp:76)
[32] /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7f4ace829d90] (??:0)
[33] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80) [0x7f4ace829e40] (??:0)
[34] /home/volter/Desktop/godot/bin/godot.linuxbsd.editor.dev.x86_64(_start+0x25) [0x563ad0a1a005] (??:?)
```

YuriSizov commented:
@MewPurPur If the issue is within the gradient itself, and not within this editor plugin, then we can go ahead with this PR (and make sure the issue is reported, so it can be looked into). Can it be reproduced without your PR? With the editor UI, or, say, by calling the reverse function from a script in a loop?

MewPurPur commented:
I don't think it's within Gradient. I'm perplexed.

MewPurPur commented:
Still no idea what it is, but I can only reproduce with one inconspicuous gradient.

```
[gd_resource type="Gradient" format=3 uid="uid://dd3hdrq23ffbj"]

[resource]
interpolation_mode = 2
offsets = PackedFloat32Array(0.1, 0.4, 0.6, 0.9)
colors = PackedColorArray(0.109667, 0, 0.14, 1, 0.63, 0.1701, 0.177765, 1, 1, 0.3825, 0.05, 1, 1, 0.978, 0.88, 1)
```

MewPurPur commented:
In spite of this, I actually can't reproduce the bug with this gradient outside of this PR.

Send help

YuriSizov commented:
I can't reproduce your bug even when holding down Undo/Redo shortcut (so things change very rapidly back and forth). There is a tiny bit of lag when the editor refreshes if you have packed arrays unfolded. But it doesn't corrupt the gradient in any way.

I say we merge it and maybe somebody will be able to reproduce your heisenbug.

MewPurPur commented:
Me neither. Hopefully it's just some sort of artifact because of this PR being up from pre-4.0 times, that won't happen in real projects. By now this is easily the weirdest bug I've ever had.

Calinou reviewed (approved):
Tested locally (rebased on top of `master` 2048fe5df5f19a5467a6fc61b9c00fe41886b134), it works as expected.

akien-mga commented:
Thanks MewPurPur for this big improvement, and thanks to all reviewers too :)
